utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/BestEffortJsonSchemaEncoder.scala

```diff
@@ -1,59 +1,52 @@
+import pl.touk.nussknacker.engine.util.json._
 
 import java.time.{LocalDate, OffsetDateTime, OffsetTime, ZonedDateTime}
 import java.util.ServiceLoader
 import scala.collection.convert.ImplicitConversions.`map AsScala`
 import scala.jdk.CollectionConverters.iterableAsScalaIterableConverter
 
+
+//TODO: Do we need validation mode here??
 class BestEffortJsonSchemaEncoder(validationMode: ValidationMode) {
 
+  import pl.touk.nussknacker.engine.util.json.JsonSchemaImplicits._
+
   private val classLoader = this.getClass.getClassLoader
   private val jsonEncoder = BestEffortJsonEncoder(failOnUnkown = false, this.getClass.getClassLoader)
 
   private val optionalEncoders = ServiceLoader.load(classOf[ToJsonBasedOnSchemaEncoder], classLoader).asScala.map(_.encoder(this.encodeBasedOnSchema))
   private val highPriority: PartialFunction[EncodeInput, EncodeOutput] = Map()
 
   final def encodeOrError(value: Any, schema: Schema): Json = {
-    encode(value, schema).valueOr(errors => throw new RuntimeException(errors.toList.mkString(",")))
+    encodeWithJsonValidation(value, schema).valueOr(errors => throw new RuntimeException(errors.toList.mkString(",")))
   }
 
   private def encodeObject(fields: Map[String, _], parentSchema: ObjectSchema): EncodeOutput = {
-    fields
-      .map(field => (field, parentSchema.getPropertySchemas.get(field._1)))
-      .collect {
-        case ((fieldName, null), schema) if isNullableSchema(schema) => Valid((fieldName, Json.Null))
-        case ((fieldName, value), propertySchema) if (propertySchema != null) && notNullOrRequired(fieldName, value, parentSchema) => encode(value, propertySchema, Some(fieldName)).map(fieldName -> _)
-        case ((fieldName, _), null) if !parentSchema.permitsAdditionalProperties() => error(s"Not expected field with name: ${fieldName} for schema: $parentSchema and policy $validationMode does not allow redundant")
-        case ((fieldName, value), null) => Option(parentSchema.getSchemaOfAdditionalProperties) match {
-          case Some(additionalPropertySchema) => encode(value, additionalPropertySchema).map(fieldName -> _)
-          case None =>  Valid(jsonEncoder.encode(value)).map(fieldName -> _)
-        }
+    fields.keys.toList.union(parentSchema.getPropertySchemas.keySet.asScala.toList).distinct.map{ key =>
+      val schema = Option(parentSchema.getPropertySchemas.get(key))
+      val value = fields.getOrElse(key, null)
+      (key, value, fields.contains(key), schema)
+    }.collect {
+      case (fieldName, null, true, Some(schema)) if schema.isNullableSchema => Valid((fieldName, Json.Null))
+      case (fieldName, null, false, Some(_)) if parentSchema.getRequiredProperties.contains(fieldName) =>
+        error(s"Missing property: $fieldName for schema: $parentSchema.")
+      case (fieldName, value, true, Some(schema)) if parentSchema.getPropertySchemas.keySet().contains(fieldName) =>
```

Comment:
I think this if is redundant, if schema is not None, it means the key is there?

Reply:
Ahh, your're right.. This if is done in patter matching..  I'm silly ;)